fuzz/src/creation.rs

```diff
@@ -0,0 +1,529 @@
+                // string on the heap
+
+                Some((compact, std_str))
+            }
+            IterChar(chars) => {
+                let compact: CompactString = chars.iter().collect();
+                let std_str: String = chars.iter().collect();
+
+                assert_eq!(compact, std_str);
+                assert_properly_allocated(&compact, &std_str);
+
+                Some((compact, std_str))
+            }
+            IterString(strings) => {
+                let compact: CompactString =
+                    strings.iter().map::<&str, _>(|s| s.as_ref()).collect();
+                let std_str: String = strings.iter().map::<&str, _>(|s| s.as_ref()).collect();
+
+                assert_eq!(compact, std_str);
+                assert_properly_allocated(&compact, &std_str);
+
+                Some((compact, std_str))
+            }
+            Bytes(data) => {
+                let compact = CompactString::from_utf8(data);
+                let std_str = std::str::from_utf8(data);
+
+                match (compact, std_str) {
+                    // valid UTF-8
+                    (Ok(c), Ok(s)) => {
+                        assert_eq!(c, s);
+                        assert_properly_allocated(&c, s);
+
+                        Some((c, s.to_string()))
+                    }
+                    // non-valid UTF-8
+                    (Err(c_err), Err(s_err)) => {
+                        assert_eq!(c_err, s_err);
+                        None
+                    }
+                    _ => panic!("CompactString and core::str read UTF-8 differently?"),
+                }
+            }
+            Buf(data) => {
+                let mut buffer = Cursor::new(data);
+
+                let compact = CompactString::from_utf8_buf(&mut buffer);
+                let std_str = std::str::from_utf8(data);
+
+                match (compact, std_str) {
+                    // valid UTF-8
+                    (Ok(c), Ok(s)) => {
+                        assert_eq!(c, s);
+                        assert_properly_allocated(&c, s);
+
+                        Some((c, s.to_string()))
+                    }
+                    // non-valid UTF-8
+                    (Err(c_err), Err(s_err)) => {
+                        assert_eq!(c_err, s_err);
+                        None
+                    }
+                    _ => panic!("CompactString and core::str read UTF-8 differently?"),
+                }
+            }
+            BufUnchecked(data) => {
+                let mut buffer = Cursor::new(data);
+
+                // The data provided might not be valid UTF-8. We mainly want to make sure we don't
+                // panic, and the data is written correctly. Before returning either of these types
+                // we'll make sure they contain valid data
+                let compact = unsafe { CompactString::from_utf8_buf_unchecked(&mut buffer) };
```

Comment:
Is this right? An exactly 24 bytes random string would be broken the result if `data[23] >= 192`, wouldn't it?

Reply:
Great observation! We [check](https://github.com/ParkMyCar/compact_str/blob/main/compact_str/src/repr/bytes.rs#L44-L55) for this exact case when reading the bytes from a `Buf`, and if we hit it, we expand to a heap string which fixes the issue.

Reply:
Ah, nice! I did not assume that you're checking the `unchecked()` data. :) But I guess it's the right way to do that, because you are allowed to build illegal Strings, and otherwise CompactString would not be a "proper" replacement.

Reply:
Valid assumption to make :) 

When I first encountered this problem I did think twice about adding the check, since the API is literally called "unchecked". But I figured it was the right thing to do because the check is very cheap, comparing a single byte on the stack, and the bug it fixes is very dangerous, possibly de-referencing a garbage pointer